k8s-kvm/provision_k8s_host_kvm.sh

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env bash
+
+# This script is intended to be run on a SLES box installed as a KVM host.
+# After OS installation, the following commands should also be run to set up directory permissions and the default network (as root):
+# usermod -aG libvirt root
+# virsh net-define net-default.xml
+# virsh net-start default
+# ntpdate time.nist.gov
+# systemctl enable ntpd
+# systemctl start ntpd
+
+set -o errexit
+
+export K8S_VM_NAME=${K8S_VM_NAME:-$1}
+export K8S_VM_IMAGE_NAME=${K8S_VM_IMAGE_NAME:-scf-libvirt-v2.0.6}
+export K8S_VM_IMAGE_PATH=${K8S_VM_IMAGE_PATH:-~/qcow2-disks}
+export K8S_VM_MEM_GIB=${K8S_VM_MEM_GIB:-8}
+
+# Used for displaying external setup information at end of provisioning
+export OBJECT_STORAGE_BASE_URL=${OBJECT_STORAGE_BASE_URL:-\${OBJECT_STORAGE_BASE_URL}}
+export OBJECT_STORAGE_YAML_BUCKET=${OBJECT_STORAGE_YAML_BUCKET:-kube-config}
+
+mkdir -p "$K8S_VM_IMAGE_PATH"
+cd "$K8S_VM_IMAGE_PATH"
+
+if [[ -z $K8S_VM_NAME ]]; then
+  echo "A VM name must be provided"
+  exit 1
+fi
+
+# Fetch base image if it doesn't exist
+if [[ ! -f $K8S_VM_IMAGE_NAME.qcow2 ]]; then
+  echo "Fetching base image:"
+  curl -L https://cf-opensusefs2.s3.amazonaws.com/vagrant/${K8S_VM_IMAGE_NAME}.box | tar --wildcards box.img -xz
+  mv box.img $K8S_VM_IMAGE_NAME.qcow2
+fi
+
+# Fetch VM access key if it doesn't exist
+if [[ ! -f vm-key ]]; then
+  curl -sL -o "vm-key" https://raw.githubusercontent.com/mitchellh/vagrant/v1.9.6/keys/vagrant
+  chmod 400 vm-key
+fi
+K8S_VM_KEY=${K8S_VM_KEY:-$PWD/vm-key} 
+
+if [[ $K8S_VM_MEM_GIB -gt $(( $(free -g | grep [0-9] | head -1 | awk '{print $2}') - 2 )) ]]; then
+  echo K8S_VM_MEM_GIB value $K8S_VM_MEM_GIB exceeds system resources
+  exit 1
+fi
+
+unset RUNNING_VM_IMAGE
+if virsh domstate $K8S_VM_NAME 2>/dev/null; then
+  echo "Deleting existing domain $K8S_VM_NAME"
+  RUNNING_VM_IMAGE=$(virsh dumpxml $K8S_VM_NAME | grep 'source file' | grep -o "'.*'" | tr -d "'")
+  if [[ -z $RUNNING_VM_IMAGE ]]; then
+    echo "Disk for domain $K8S_VM_NAME could not be determined. Please delete this domain manually"
+    exit 1
+  fi
+  virsh destroy "$K8S_VM_NAME"
+  virsh undefine "$K8S_VM_NAME"
+  rm -f "$RUNNING_VM_IMAGE"
+fi
+
+cp "$K8S_VM_IMAGE_NAME.qcow2" "$K8S_VM_IMAGE_NAME-$K8S_VM_NAME.qcow2"
```

Comment:
This seems to be missing.  It's the difference between ~instant and ~minutes, so please have it 😁

Reply:
This tends to take ~15s for me IIRC, though I don't doubt it could take minutes depending on your disk/FS. Even with `reflink=auto`.. isn't the actual filesystem copy going to occur as soon as the VM is started? If I understand correctly, that just means we're postponing this overhead for one command. With the copy already behind us, I haven't seen it take more than 30 seconds to get the IP from the virtual network (though I don't doubt it can happen).. see below where I suggest a 1 minute timeout.. maybe that needs to be increased as well? The problem is, I have no way of knowing what a reasonable timeout there is. Especially if the actual on-disk copy is happening at that point, couldn't that take upwards of 5 minutes for users with slow disks?

Reply:
You'd normally not have to write every cluster of the image (most of it is read-only data, things like binaries). So only changes would need to be copied.